sql/core/src/main/scala/org/apache/spark/sql/Diff.scala

```diff
@@ -0,0 +1,155 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import org.apache.spark.sql.functions.{coalesce, lit, not, when}
+
+/**
+ * Differ class to diff two Datasets.
+ * @param options options for the diffing process
+ */
+class Diff(options: DiffOptions) {
+
+  def checkSchema[T](left: Dataset[T], right: Dataset[T], idColumns: String*): Unit = {
+    require(left.columns.length == right.columns.length,
+      "The number of columns doesn't match.\n" +
+        s"Left column names (${left.columns.length}): ${left.columns.mkString(", ")}\n" +
+        s"Right column names (${right.columns.length}): ${right.columns.mkString(", ")}")
+
+    require(left.columns.length > 0, "The schema must not be empty")
+
+    // we ignore the nullability of fields
+    val leftFields = left.schema.fields.map(f => (f.name, f.dataType))
+    val rightFields = right.schema.fields.map(f => (f.name, f.dataType))
+    val leftExtraSchema = leftFields.diff(rightFields)
+    val rightExtraSchema = rightFields.diff(leftFields)
+    val extraSchema = leftExtraSchema.union(rightExtraSchema)
+    require(extraSchema.isEmpty,
```

Comment:
I presume this will flag  up differences if the case of the column names are different.  This is different to how spark generally treats this, as normally column names are case insensitive. Is the behaviour here correct (arguably it is) and can we document somewhere.

Reply:
My other thought was that this sort of code (checking two schemata for equality) must exist somewhere else in spark?  Is there any code we could reuse or conventions we could follow?

Reply:
Column name case sensitivity is configurable, added respective code. Regarding existing schema equality: I could not find any, spark reviewers might point me to it.